### PR TITLE
fix Eval sync issue

### PR DESCRIPTION
**Pull request by yxxhero**
Signed-off-by: yxxhero <aiopsclub@163.com>

fix Eval sync issue

fix: https://github.com/helmfile/helmfile/issues/429

### CODE REVIEW

**Comment by yxxhero** [comment]
@mumoshu

File: vals.go [hunk 1]
```diff
@@ -224,6 +227,9 @@ func (r *Runtime) Eval(template map[string]interface{}) (map[string]interface{},
 			}
 
 			hash := uriToProviderHash(uri)
+
+			r.m.Lock()
+			defer r.m.Unlock()
```
Comment (mumoshu) [inline vals.go]: Thanks!
This is going to retain the lock until L331, which seems to too long at glance.
Perhaps we can make it more coerce, by wrapping the critical section using the additional block, or extracting the whole block along with lock/defer-unlock into another function?

For example:

```
		{
			r.m.Lock()
			defer r.m.Unlock()

			p, ok := r.providers[hash]
			if !ok {
				var scheme string
				scheme = uri.Scheme
				scheme = strings.Split(scheme, "://")[0]
				p, err = createProvider(scheme, uri)
				if err != nil {
					return "", err
				}
				r.providers[hash] = p
			}
		}
```

Comment (yxxhero) [reply vals.go]: @mumoshu done. WDYT? https://github.com/variantdev/vals/pull/104/commits/84daedc40d03ad3b2fd003578966e71bc7f4e197

**Comment by yxxhero** [comment]
@mumoshu done

**Review by mumoshu — APPROVED** [review]
LGTM. Thanks a lot for fixing it!

**Comment by yxxhero** [comment]
@mumoshu I think should use squash and merge. Or I squash commits locally.

**Comment by pathob** [comment]
Thank you @yxxhero and @mumoshu 🙏

**Pull request merged**